### PR TITLE
Develop ts_group_by_key

### PR DESCRIPTION
**Pull request by sergey-lipikhin**

### CODE REVIEW

**Review by vladyslav-mosor — APPROVED** [review]
GJ using reduce)

File: src/groupByKey.ts [hunk 1]
```diff
@@ -2,6 +2,19 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+export function groupByKey<T>(items: T[], key: keyof T): object {
```
Comment (sergii-nosachenko) [inline src/groupByKey.ts:5]: The result is not just any object. Use `GroupsMap<T>` instead.

File: src/groupByKey.ts [hunk 2]
```diff
@@ -2,6 +2,19 @@ type GroupsMap<T> = {
...
+  return items.reduce((
+    groupOfObejcts: GroupsMap<T>,
+    singleObject,
+  ) => {
+    const groupKey = singleObject[key] as string;
```
Comment (sergii-nosachenko) [inline src/groupByKey.ts:10]: Avoid type casting. Just convert the value to a string.

File: src/groupByKey.ts [hunk 3]
```diff
@@ -2,6 +2,19 @@ type GroupsMap<T> = {
...
+
+    if (!Object.prototype.hasOwnProperty.call(groupOfObejcts, groupKey)) {
+      Object.assign(groupOfObejcts, { [groupKey]: [] });
+    }
```
Comment (sergii-nosachenko) [inline src/groupByKey.ts:14]: Use advanced techniques:
```suggestion
    groupOfObejcts[groupKey] ??= [];
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment

File: src/groupByKey.ts [hunk 4]
```diff
@@ -2,6 +2,19 @@ type GroupsMap<T> = {
...
+
+    groupOfObejcts[groupKey].push(singleObject);
+
+    return groupOfObejcts;
+  }, {}) as object;
```
Comment (sergii-nosachenko) [inline src/groupByKey.ts:19]: Fix unnecessary type casting.

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]

**Review by Yevhenii0536 — APPROVED** [review]
why did you made so hardreading code? :D try to make simple

**Review by OleksiiBlezniuk — APPROVED** [review]

**Pull request closed without merging**